File: backend-node/src/routes/webSocketRoutes.ts
import { Server } from 'node:http';

import { WebSocketServer } from 'ws';
import type { WebSocket } from 'ws';
import { v4 as uuid } from 'uuid';

import { ServiceContainer } from '../container/ServiceContainer';
import { logger } from '../utils/logger';

export function registerWebSocketRoutes(server: Server, container: ServiceContainer): void {
  const webSocketManager = container.getWebSocketManager();
  const streamingService = container.getStreamingService();

  const wss = new WebSocketServer({ server, path: '/api/playback/stream' });

  wss.on('connection', async (socket: WebSocket, req) => {
    const sessionId = uuid();
    const url = new URL(req.url ?? '', `http://${req.headers.host}`);
    const roomId = url.searchParams.get('roomId');

    if (!roomId) {
      logger.error({ sessionId }, 'WebSocket connection missing roomId');
      socket.close(1008, 'roomId required');
      return;
    }

    logger.info({ sessionId, roomId }, 'WebSocket client connected');
    webSocketManager.addConnection(roomId, sessionId, socket);

    try {
      const state = await streamingService.getState(roomId);
      socket.send(JSON.stringify(state));
    } catch (error) {
      logger.error({ error, sessionId, roomId }, 'Failed to send initial state');
    }

    socket.on('message', async (message: Buffer) => {
      const payload = message.toString();
      if (payload === 'ping') {
        socket.send('pong');
        return;
      }

      try {
        const data = JSON.parse(payload);
        if (data.type === 'player_tick' && data.payload) {
          await webSocketManager.handlePlayerTick(roomId, data.payload);
        }
      } catch (error) {
        // logger.debug({ error, sessionId }, 'Failed to parse WebSocket message');
      }
    });

    socket.on('close', () => {
      webSocketManager.removeConnection(roomId, sessionId);
      logger.info({ sessionId, roomId }, 'WebSocket client disconnected');
    });

    socket.on('error', (error: Error) => {
      logger.error({ error, sessionId, roomId }, 'WebSocket error');
      webSocketManager.removeConnection(roomId, sessionId);
    });
  });
}

File: backend/src/main/kotlin/com/musicstreaming/api/WebSocketRoutes.kt
package com.musicstreaming.api

import com.musicstreaming.di.ServiceContainer
import io.ktor.server.routing.*
import io.ktor.server.websocket.*
import io.ktor.websocket.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.slf4j.LoggerFactory
import java.util.*

private val logger = LoggerFactory.getLogger("WebSocketRoutes")

/**
 * Configure WebSocket routes for real-time playback state updates
 * Requirement 8.4: WebSocket endpoint for real-time updates
 * Requirement 9.4: Connection management
 */
fun Route.webSocketRoutes() {
    val webSocketManager = ServiceContainer.getInstance().webSocketManager
    val streamingService = ServiceContainer.getInstance().streamingService
    val json = Json {
        prettyPrint = false
        ignoreUnknownKeys = true
    }
    
    // WS /api/playback/stream - Real-time playback state updates
    webSocket("/api/playback/stream") {
        val sessionId = UUID.randomUUID().toString()
        
        try {
            // Register the connection
            webSocketManager.addConnection(sessionId, this)
            logger.info("WebSocket client connected: $sessionId")
            
            // Send initial state to the newly connected client
            try {
                val currentState = streamingService.getState()
                val stateJson = json.encodeToString(currentState)
                send(Frame.Text(stateJson))
                logger.info("Initial state sent to client: $sessionId")
            } catch (e: Exception) {
                logger.error("Failed to send initial state to client $sessionId: ${e.message}", e)
            }
            
            // Keep the connection alive and handle incoming messages
            for (frame in incoming) {
                when (frame) {
                    is Frame.Text -> {
                        val text = frame.readText()
                        logger.debug("Received message from client $sessionId: $text")
                        
                        // Handle ping/pong or other client messages if needed
                        if (text == "ping") {
                            send(Frame.Text("pong"))
                        }
                    }
                    is Frame.Close -> {
                        logger.info("Client $sessionId requested close")
                        break
                    }
                    else -> {
                        logger.debug("Received non-text frame from client $sessionId")
                    }
                }
            }
        } catch (e: Exception) {
            logger.error("WebSocket error for client $sessionId: ${e.message}", e)
        } finally {
            // Clean up the connection
            webSocketManager.removeConnection(sessionId)
            logger.info("WebSocket client disconnected: $sessionId")
        }
    }
}
